delegate: simplify / speed up SHOW FUNCTIONS; redefine several pg builtins as UDFs; fix pg_proc virtual index

jordanlewis opened the pull request:
Epic: CRDB-23454
Closes #94746
Closes #94953

Remove a couple of unnecessary builtin function calls in SHOW FUNCTIONS.
The query used to take 50 seconds, now it takes less than a second.

Also, rewrite several pg compatibility builtins as UDFs for improved performance.

Also, fix the pg_proc oid virtual index which didn't work for user-defined functions.

Epic: None

jordanlewis committed:
delegate: simplify / speed up SHOW FUNCTIONS

Remove a couple of unnecessary builtin function calls in SHOW FUNCTIONS.
The query used to take 50 seconds, now it takes less than a second.

Release note (performance improvement): improve the performance of SHOW
FUNCTIONS

jordanlewis committed:
builtins: rewrite several builtins as UDFs

Release note (performance improvement): several pg compatibility
builtins have improved performance

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/94771)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94771)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[`pkg/sql/delegate/show_functions.go` line 75 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NL2XXl68K4t3NXa1sCb:-NL2XXl68K4t3NXa1sCc:bazm1d5) ([raw file](https://github.com/cockroachdb/cockroach/blob/cad0a6b2d637223b3be3dd04ffceb160bb5a9a9b/pkg/sql/delegate/show_functions.go#L75)):*
> ```Go
> WHERE true
> %[2]s
> GROUP BY n.nspname, p.proname, p.prorettype, p.prokind, p.provolatile
> ```

why do we need the GROUP BY? also, could we just do `group by 1, 2, 3, 4`
___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 274 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NL2YP2pCzfEEghehorx:-NL2YP2pCzfEEghehory:br5ce8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/13cf0f9b16e6a4b27f21b4cf226ad81a3420e192/pkg/sql/sem/builtins/pg_builtins.go#L274)):*
> ```Go
> 		FROM pg_catalog.pg_matviews v
> 		JOIN pg_catalog.pg_class c ON c.relname=v.matviewname
> 		WHERE c.oid=$1`,
> ```

looks like  this should return NULL when there's no view with that OID. does this new version do that, or return 0 rows? (and is it already tested?)
___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 287 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NL2Y1IHAmhDviNFDLu8:-NL2Y1II8SFcMSqE9SNI:b-1b92n3) ([raw file](https://github.com/cockroachdb/cockroach/blob/13cf0f9b16e6a4b27f21b4cf226ad81a3420e192/pkg/sql/sem/builtins/pg_builtins.go#L287)):*
> ```Go
> 		IsUDF:      true,
> 		Body: `SELECT COALESCE((SELECT condef FROM pg_catalog.pg_constraint WHERE oid=$1),
> 													  'unknown constraint (OID=' || $1 || ')')`,
> ```

is this right? instead of returning an error, this will return a string row.

also, it seems like the PG behavior is that null is returned in this case, so i'm not sure why the old code returned a pgerror
___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 653 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NL2ZXaN6JEb_Ls7uIGu:-NL2ZXaN6JEb_Ls7uIGv:b-dzjeuh) ([raw file](https://github.com/cockroachdb/cockroach/blob/13cf0f9b16e6a4b27f21b4cf226ad81a3420e192/pkg/sql/sem/builtins/pg_builtins.go#L653)):*
> ```Go
> 			ReturnType: tree.FixedReturnType(types.String),
> 			IsUDF:      true,
> 			Body: `SELECT trim('{}' FROM
> ```

nit: could you a comment saying how it's getting reformatted (like you added below)
___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 873 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NL2_lku0MUfvklk1_90:-NL2_lku0MUfvklk1_91:b-ve51o6) ([raw file](https://github.com/cockroachdb/cockroach/blob/13cf0f9b16e6a4b27f21b4cf226ad81a3420e192/pkg/sql/sem/builtins/pg_builtins.go#L873)):*
> ```Go
> 		tree.Overload{
> 			Types:      tree.ParamTypes{{Name: "sequence_oid", Typ: types.Oid}},
> 			ReturnType: tree.FixedReturnType(types.String),
> ```

can the return type change to a record now?
___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 877 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NL2_PcE5HFwqD4oQNCp:-NL2_PcE5HFwqD4oQNCq:byrxdsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/13cf0f9b16e6a4b27f21b4cf226ad81a3420e192/pkg/sql/sem/builtins/pg_builtins.go#L877)):*
> ```Go
> 			Body: `SELECT COALESCE (((seqstart, seqmin, seqmax, seqincrement, seqcycle, seqcache, seqtypeid)
>              FROM pg_catalog.pg_sequence WHERE seqrelid=$1 LIMIT 1),
>              'unknown sequence (OID=' || $1 || ')')`,
> ```

hm in PG this one actually returns an error, but now it will just return this message in the results

crazy how the `pg_*` builtins are inconsistent about how they behave with not found entities


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94771)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/delegate/show_functions.go` line 75 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NL2XXl68K4t3NXa1sCb:-NL2jZ836RokxC_9Z_JR:buve60j) ([raw file](https://github.com/cockroachdb/cockroach/blob/cad0a6b2d637223b3be3dd04ffceb160bb5a9a9b/pkg/sql/delegate/show_functions.go#L75)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why do we need the GROUP BY? also, could we just do `group by 1, 2, 3, 4`
</blockquote></details>

The presence of the `array_agg` requires the group by. Okay, that's a better pattern.
___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 274 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NL2YP2pCzfEEghehorx:-NL2jcVv-M4Dl91-AGSE:b-uogmq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/13cf0f9b16e6a4b27f21b4cf226ad81a3420e192/pkg/sql/sem/builtins/pg_builtins.go#L274)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

looks like  this should return NULL when there's no view with that OID. does this new version do that, or return 0 rows? (and is it already tested?)
</blockquote></details>

Returning 0 rows here is the same as returning NULL - all of the builtin functions like this one behave like that
___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 287 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NL2Y1IHAmhDviNFDLu8:-NL2qWGv1uqKkxnHQeLl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13cf0f9b16e6a4b27f21b4cf226ad81a3420e192/pkg/sql/sem/builtins/pg_builtins.go#L287)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is this right? instead of returning an error, this will return a string row.

also, it seems like the PG behavior is that null is returned in this case, so i'm not sure why the old code returned a pgerror
</blockquote></details>

Done.
___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 873 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NL2_lku0MUfvklk1_90:-NL38pPL7jYzpL3YRCLl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13cf0f9b16e6a4b27f21b4cf226ad81a3420e192/pkg/sql/sem/builtins/pg_builtins.go#L873)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can the return type change to a record now?
</blockquote></details>

Done.
___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 877 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NL2_PcE5HFwqD4oQNCp:-NL34SSQASk424ObONr2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13cf0f9b16e6a4b27f21b4cf226ad81a3420e192/pkg/sql/sem/builtins/pg_builtins.go#L877)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm in PG this one actually returns an error, but now it will just return this message in the results

crazy how the `pg_*` builtins are inconsistent about how they behave with not found entities
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jordanlewis committed:
sql: fix pg_proc virtual oid index

Previously, the pg_proc virtual oid index was broken for user-defined
functions.

Release note: None (no release with this bug)

jordanlewis commented:
Lots of little things to fix in there - anyway it's RFAL.

rafiss reviewed (approved):
cool!

> Reviewed 1 of 2 files at r3, 2 of 3 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-:-NLM8VOgCOnPzMlDY9oD:b-7s4s8d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rharding6373)

___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 1109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NLM8nyw1FejXYy3tAAZ:-NLM8nyw1FejXYy3tAA_:bd6g8sl) ([raw file](https://github.com/cockroachdb/cockroach/blob/42b0487ffc333e713a169ec121c576bfdb2df659/pkg/sql/sem/builtins/pg_builtins.go#L1109)):*
> ```Go
> 	),
> 
> 	"obj_description": makeBuiltin(defProps(),
> ```

obj_description and col_description are also great candidates for this rewrite. we've seen both come up with performance issues in the past


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94771)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rharding6373)

___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 1109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NLM8nyw1FejXYy3tAAZ:-NLM9QXgCeZ-lrv4uqMJ:b-l3ycgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/42b0487ffc333e713a169ec121c576bfdb2df659/pkg/sql/sem/builtins/pg_builtins.go#L1109)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

obj_description and col_description are also great candidates for this rewrite. we've seen both come up with performance issues in the past
</blockquote></details>

https://github.com/cockroachdb/cockroach/issues/87947


<!-- Sent from Reviewable.io -->

rharding6373 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but @mgartner  is the UDF expert so I've added him to take a look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-:-NLNDPUL-wqk4NSiRy9H:b74pylp)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @mgartner, and @rafiss)

___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 887 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NLNDwTM79pZtx_oqXMj:-NLNDwTM79pZtx_oqXMk:b-f4fg6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4027a0ab8d35724942d4aa4c24b97798faba22d/pkg/sql/sem/builtins/pg_builtins.go#L887)):*
> ```Go
> 		// The real implementation returns a record; we fake it by returning a
> 		// comma-delimited string enclosed by parentheses.
> 		// TODO(jordan): convert this to return a record type once we support that.
> ```

nit: is this TODO obsolete now?
___
*[`pkg/sql/sem/builtins/pg_builtins.go` line 1003 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/94771#-NLNEFcyEMuPT37RHNAR:-NLNEFcyEMuPT37RHNAS:blo39kg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4027a0ab8d35724942d4aa4c24b97798faba22d/pkg/sql/sem/builtins/pg_builtins.go#L1003)):*
> ```Go
> 			FROM pg_catalog.pg_description
> 			WHERE objoid=$1
> 		  AND objsubid = 0
> ```

nit: this body and the next look like they have odd whitespace formatting


<!-- Sent from Reviewable.io -->

rafiss reviewed (dismissed):
assuming this fixes https://github.com/cockroachdb/cockroach/issues/94953 could you add a test in `pkg/sql/logictest/testdata/logic_test/pg_builtins` for `pg_function_is_visible`?  see the analagous tests for `pg_type_is_visible`

mgartner commented on pkg/sql/logictest/testdata/logic_test/udf:
```diff
@@ -174,6 +174,13 @@ FROM pg_catalog.pg_proc WHERE proname IN ('proc_f', 'proc_f_2');
+# Ensure that the pg_proc virtual index works properly.
+
+query TT
+SELECT oid, proname FROM pg_proc WHERE oid = 'sc.proc_f_2'::regproc
```
nit: force the index with `pg_proc@pg_proc_oid_idx`

jordanlewis replied:
You can't force virtual indexes - the optimizer gives an error.

mgartner reviewed (approved):
Cool! LGTM

jordanlewis commented:
bors r+
